File: src/patterns/behavioral/CoR/CoR.go
package main

import (
	"errors"
)

type RequestOptions struct {
	isAuthenticated bool
	isValid bool
	isBanned bool
}

// Handler

type IRequestHandler interface {
	SetNext(handler IRequestHandler) IRequestHandler
	Handle(options *RequestOptions) (string, error)
}

// BaseHandler

type BaseHandler struct {
	next IRequestHandler
}

func (b *BaseHandler) SetNext(handler IRequestHandler) IRequestHandler { 
	b.next = handler
	return b.next
}

func (b *BaseHandler) Handle(options *RequestOptions) (string, error) {
	var err error;

	if b.next != nil {
		return b.Handle(options)
	}

	err = errors.New("there is not next handler available")

	return "", err
}

// ConcreteHandlers

type AuthHandler struct {
	BaseHandler
}

func (a *AuthHandler) Handle(options *RequestOptions) (string, error) {
	var err error;

	if a.next != nil && options.isAuthenticated == true {
		return a.next.Handle(options)
	}

	err = errors.New("user not authenticated")
	return "", err
}

type ValidationHandler struct {
	BaseHandler
}

func (v *ValidationHandler) Handle(options *RequestOptions) (string, error) {
	var err error;

	if v.next != nil && options.isValid == true {
		return v.next.Handle(options)
	}

	err = errors.New("request is not valid")
	return "", err
}

type BanHandler struct {
	BaseHandler
}

func (b *BanHandler) Handle(options *RequestOptions) (string, error) {
	var err error;

	if b.next != nil && options.isBanned == false {
		return b.next.Handle(options)
	}

	err = errors.New("user is banned")
	return "", err
}

type RouteHandler struct {
	BaseHandler
}

func (r *RouteHandler) Handle(options *RequestOptions) (string, error) {
	return "User is authenticated successfully", nil
}

// func main() {
// 	userOptions := &RequestOptions{isAuthenticated: false, isBanned:true, isValid: true}

// 	authHandler := &AuthHandler{}
// 	validationHandler := &ValidationHandler{}
// 	bannedHandler := &BanHandler{}	
// 	routeHandler := &RouteHandler{}

// 	bannedHandler.SetNext(validationHandler).SetNext(authHandler).SetNext(routeHandler)

// 	fmt.Println(bannedHandler.Handle(userOptions))
// }

File: src/patterns/behavioral/CoR/CoR.ts
class RequestClass {
  isAuthenticated: boolean;
  isValid: boolean;
  isBanned: boolean;

  constructor({
    isAuthenticated,
    isValid,
    isBanned,
  }: {
    isAuthenticated: boolean;
    isValid: boolean;
    isBanned: boolean;
  }) {
    this.isAuthenticated = isAuthenticated;
    this.isValid = isValid;
    this.isBanned = isBanned;
  }
}

// Handler

interface IRequestHandler {
  setNext(handler: IRequestHandler): IRequestHandler;
  handle(options: RequestClass): string | null;
}

// BaseHandler

abstract class BaseRequestHandler implements IRequestHandler {
  next: IRequestHandler | null = null;

  setNext(handler: IRequestHandler): IRequestHandler {
    this.next = handler;
    return this.next;
  }

  handle(options: RequestClass): string | null {
    if (this.next) {
      this.next.handle(options);
    }

    return null;
  }
}

// ConcreteHandlers

class AuthenticationHandler extends BaseRequestHandler {
  handle(options: RequestClass): string | null {
    if (options.isAuthenticated && this.next !== null) {
      return this.next.handle(options);
    }

    return "User not authenticated";
  }
}

class ValidationHandler extends BaseRequestHandler {
  handle(options: RequestClass): string | null {
    if (options.isValid && this.next !== null) {
      return this.next.handle(options);
    }

    return "Request is not valid";
  }
}

class BanHandler extends BaseRequestHandler {
  handle(options: RequestClass): string | null {
    if (!options.isBanned && this.next !== null) {
      return this.next.handle(options);
    }

    return "User is banned";
  }
}

class RouteHandler extends BaseRequestHandler {
  handle(options: RequestClass): string | null {
    return "User authenticated";
  }
}

const user1Options = new RequestClass({
  isBanned: false,
  isValid: true,
  isAuthenticated: true,
});

const banHandler = new BanHandler();
const validationHandler = new ValidationHandler();
const authHandler = new AuthenticationHandler();
const routeHandler = new RouteHandler();

banHandler
  .setNext(validationHandler)
  .setNext(authHandler)
  .setNext(routeHandler);

console.log(banHandler.handle(user1Options));
